### DISCUSSION TITLE
Support for search endpoint /search/query

### DISCUSSION THREAD

**Discussion by smaar** [O365/python-o365#1092] [Ideas]
First off, great job with this library. Enjoying working with it so far and I am happy to extend it for my needs where required. Before I start extending it though I wanted to check if you already plan to add support for the /search/query endpoint. 

Reasoning is that the drive search is unfortunately a bit limited in Microsofts side. It does not return all files that I can actually access, but only those that are in my Drive or that are actively shared with me - that does for some reason not include files in Sharepoint sites that I have access to, even though in theory they are shared with me.

So to really search everywhere I will need to sue the /search/query endpoint

**Comment by alejcas** [comment]
Search is implemented. Look into the code under drive.py

**Reply by smaar** [reply]
thanks. However that is search on the drive resource like /drive/search, drive/root/search or me/drive/search.

What I am proposing is the /search/query which works on a different resource. 
<img width="948" alt="image" src="https://github.com/O365/python-o365/assets/24751614/2fe1edeb-03e0-4541-a049-6a45ee54947b">

**Reply by alejcas** [reply]
I think it works with any item. Just use
```python
drive = Storage.get_default_drive(request_drive=False)
items = drive.search()
```

if this is not working pls report back

**Reply by smaar** [reply]
its working, but it just brings me back to what I am reporting. This approach only searches in the default drive, not in any of the other drives that I have access to implicitly through my access to different Sharepoint sites. The search/query endpoint on the other hand allows you to do exactly that. 

Even first doing get_drives and then iterating through all the drives would not work because it still does not return ALL the drives that I have access to.

**Reply by smaar** [reply]
and if I use the Storage object directly, I am getting following error. I think I just use it wrong usually I do it with the account.storage() way

TypeError: Storage.get_default_drive() missing 1 required positional argument: 'self'

**Reply by alejcas** [reply]
Then I'm sorry but it's not implemented. PR are welcome.
Thanks

**Reply by smaar** [reply]
No problem, I'll look into it. Could you let me know where to best add that capability? It seems to me that its best to create a new class for this (similar to Drive, Sharepoint etc), since "search" is its own resource from Graph API perspective. What do you think?

**Reply by alejcas** [reply]
But it will search only in document libraries right?

If yes should be in drive.py (sharepoint.py uses drive.py)

**Reply by smaar** [reply]
I was able to serve my use case now with a mix of the existing search functionality and the get_shared_with_me() function.


        `drive = self.get_drive(drive_id)
        items = drive.search(".xlsx")
        shared_items = drive.get_shared_with_me(allow_external=True)
        workbooks = [
            {"id": item.object_id, "name": item.name, "drive_id": item.drive_id}
            for item in itertools.chain(items, shared_items)
            if item.is_file and item.name.endswith(".xlsx")
        ]
        return workbooks`

So I wont work on that PR for now
